Github Action to lock and unlock a service (#948)
port-labs/port-docs: docs/create-self-service-experiences/setup-backend/github-workflow/examples/lock-and_unlock-service-in-port.md

---
sidebar_position: 9
---

import PortTooltip from "/src/components/tooltip/tooltip.jsx";

# Lock and Unlock Service

> In this guide, we will create a self-service action in Port that executes a GitHub workflow to lock and unlock a service. The locking and unlocking mechanism involves updating the `locked` entity property in Port on a `production` or `development` environment. The action finally sends a Slack notification to a channel.

Reviewer: ```suggestion
In this guide, we will create a self-service action in Port that executes a GitHub workflow to lock and unlock a service. The locking and unlocking mechanism involves updating the `locked` property in Port on a `production` or `development` environment. Additionally, the action will send a Slack notification to a designated channel.
```

## Prerequisites

> 1. Install Port's GitHub app by clicking [here](https://github.com/apps/getport-io/installations/new)

Reviewer: ```suggestion
1. Install Port's GitHub app by clicking [here](https://github.com/apps/getport-io/installations/new).
```

> 2. Configure a Slack app that can post a message to a Slack channel. Tha app should have a `chat:write` bot scope under **OAuth & Permissions**

Reviewer: ```suggestion
2. Configure a [Slack app](https://api.slack.com/apps) that can post a message to a Slack channel. The app should have a `chat:write` bot scope under **OAuth & Permissions**.
```

> 3. GitHub repository in which you can trigger a workflow that we will use in this guide

Reviewer: ```suggestion
3. A GitHub repository in which you can trigger a workflow that we will use in this guide.
```

          "Library": "green"
        },
        "icon": "DefaultProperty"
      },
      "lifecycle": {
        "title": "Lifecycle",
        "type": "string",
        "enum": [
          "Production",
          "Staging",
          "Development"
        ],
        "enumColors": {
          "Production": "green",
          "Staging": "yellow",
          "Development": "blue"
        },
        "icon": "DefaultProperty"
      },
      "locked_in_prod": {
        "icon": "DefaultProperty",
        "title": "Locked in Prod",
        "type": "boolean",
        "default": false
      },
      "locked_reason_prod": {
        "icon": "DefaultProperty",
        "title": "Locked Reason Prod",
        "type": "string"
      },
      "locked_in_test": {
        "icon": "DefaultProperty",
        "title": "Locked in Test",
        "type": "boolean",
        "default": false
      },
      "locked_reason_test": {
        "icon": "DefaultProperty",
        "title": "Locked Reason Test",
        "type": "string"
      }
    },
    "required": []
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {}
}
```
</details>

:::note Lock property

> Our Service blueprint has a boolean property called `locked_in_prod`. When the action is triggered, we will update the value of this field to `true` if locking and `false` if unlocking. The `locked_reason_prod` property will equally be updated to reflect the reason for the lock or unlock. The same logic applies to the development environment (`locked_in_test`)

Reviewer: ```suggestion
Our Service blueprint has a boolean property called `locked_in_prod`. When the action is triggered, we will update the value of this field to `true` if locking and `false` if unlocking. The `locked_reason_prod` property will be updated to reflect the reason for the lock or unlock. The same logic applies to the development environment (`locked_in_test`).
```

- `<GITHUB-REPO-NAME>` - your GitHub repository name.
:::

```json showLineNumbers
[
{
  "identifier": "unlock_service",
  "title": "Unlock Service",
  "icon": "Unlock",
  "userInputs": {
    "properties": {
      "reason": {
        "title": "Reason",
        "type": "string"
      },
      "environment": {
        "icon": "DefaultProperty",
        "title": "Environment",
        "type": "string",
        "enum": [
          "Production",
          "Development",
          "Staging"
        ],
        "enumColors": {
          "Production": "lightGray",
          "Development": "lightGray",
          "Staging": "lightGray"
        }
      }
    },
    "required": [
      "environment",
      "reason"
    ],
    "order": [
      "environment",
      "reason"
    ]
  },
  "invocationMethod": {
    "type": "GITHUB",
    "org": "<GITHUB-ORG>",
    "repo": "<GITHUB-REPO-NAME>",
    "workflow": "unlock_service.yaml",
    "omitUserInputs": false,
    "omitPayload": false,
    "reportWorkflowStatus": true
  },
  "trigger": "DAY-2",
  "description": "Unlock service in Port",
  "requiredApproval": false
}
]
```

</details>
<br />

> 3. Create a workflow file under `.github/workflows/lock_service.yml` with the following content:

Reviewer: ```suggestion
3. In your Github repository, create a workflow file under `.github/workflows/lock_service.yml` with the following content:
```

          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          runId: ${{fromJson(github.event.inputs.port_payload).context.runId}}
          logMessage: "The unlock operation has been completed successfully and the details is being broadcasted to Slack. The outcome of the Slack announcement is ${{ steps.slack.outcome }}"
          
      - name: Inform unsuccessful service unlocking in Port
        if: ${{ (steps.unlock-prod-service.outcome != 'success' && steps.unlock-prod-service.outcome != 'skipped') || (steps.unlock-test-service.outcome != 'success' && steps.unlock-test-service.outcome != 'skipped') }}
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          runId: ${{fromJson(inputs.port_payload).context.runId}}
          logMessage: The attempt to unlock the service was not successful
```

</details>
<br />

5. Trigger the actions from the [self-service](https://app.getport.io/self-serve) page of your Port application.

> Below is the result of a succesful service lock and unlock alert sent to a Slack channel

Reviewer: ```suggestion
Below is the result of a successful service lock and unlock alert sent to a Slack channel:
```
